### ISSUE TITLE
Misc. bug: backend sampler probe hardcodes a 1M-row width, turning ARGSORT's fallback check into a hard gate

### ISSUE THREAD

**Issue by dlsniper** [ggml-org/llama.cpp#28513]
### Name and Version

```
$ ./llama-cli --version
version: 0.3.0-dev (build 10798, commit c390d0abb)
built with GNU 13.3.0 for Linux x86_64
```

Also reproduced on the `b10819` release tarball. `src/llama-sampler.cpp` is
byte-identical between `b10798` and master.

### Operating systems

Linux

### Which llama.cpp modules do you know to be affected?

libllama (core library), Test code

### Command line

```shell
./repro Qwen3-0.6B-Q8_0.gguf -ngl 99 -chain 4
```

### Problem description & steps to reproduce

gfx1151 (Radeon 8060S, Ryzen AI MAX+ 395), ROCm 10.0, Linux x64. Qwen3-0.6B-Q8_0, `n_vocab` 151936. Reproducer below, `llama.h` only.

- `llama_sampler_backend_support()` (`src/llama-sampler.cpp:638`) runs per-sampler, from each `backend_init` (`:1470` `:1620` `:1819` `:2220` `:1263`).
- It probes a hardcoded `1024*1024`-row logits tensor (`:583` `:603` `:647` `:767`) — never the width the sampler receives; no shape is threaded between chain entries.
- HIP: `top_k(40) -> top_p` is refused over an ARGSORT of 40 rows, a width HIP supports.
- The refusal drops the whole offload suffix, `dist` included (`:745-765`).
- That ARGSORT cap exists to enable the CPU fallback — #16323: "check so we can fall back to CPU".
- `llama_set_sampler()` returns `true` regardless: `can_offload` from pointer presence, `backend_init()` at `src/llama-context.cpp:1247`, `bool` discarded, `return true` `:1253`.
- Unreachable for that case: the `cannot be offloaded to the backend` warning (`:1257`), and `common/speculative.cpp:504,1036,1405`.

Chain ablation on HIP:

| chain                   | cand_count | probs_count | backend token | stderr  |
|-------------------------|------------|-------------|---------------|---------|
| `dist`                  | 0          | 151936      | valid         | —       |
| `temp,dist`             | 0          | 151936      | valid         | —       |
| `temp,top_k,dist`       | 40         | 40          | valid         | —       |
| `temp,top_k,top_p,dist` | 40         | **0**       | **-1**        | ARGSORT |
| `temp,top_p,dist`       | **0**      | **0**       | **-1**        | ARGSORT |

Probe bypassed (`:647,767`, `1024*1024` -> `1024`), `libllama.so` rebuilt against the release `libggml-hip.so`, unpatched control confirms stock behaviour:

| chain                   | stock         | probe bypassed                     |
|-------------------------|---------------|------------------------------------|
| `temp,top_k,top_p,dist` | `-1`, probs 0 | token 11, cand 40, probs 40        |
| `temp,top_p,dist`       | `-1`, probs 0 | token 11, cand/probs 151936, sum 1 |

- Row 2 argsorts 151936 rows: `ggml_backend_sched` split that node to CPU (`graph splits = 3` vs control `1`) and returned correct normalized probabilities.
- Ack: split-avoidance is a stated goal (#17004); per-node fallback has a measured cost (#27865, ~50% decode at 42k ctx). Both are per layer per token; the sampler tail runs once per token.
- Ack: partial offload is intended (#23287); `-1`/NULL are documented sentinels (`include/llama.h:39,1074,1079`). The gap is a configuration-time signal, not read-time detection.
- Not asking for CUB on HIP; that is #26592.

Separately, `tests/test-backend-sampler.cpp:2105-2114` skips six tests on HIP (TODO'd against #26592):

- `set_sampler` (`:1449`, chain `dist`) and `mixed` (`:1399`, chains `dist` and `top_k(40)`) build no `ggml_argsort`, and TOP_K is HIP-supported since #27466 — both skips look stale. Reproducer runs both shapes (`-chain 1`, `-chain 3`).
- Skip message reads `no backend TOP_K support`; for the still-justified skips the blocking op is ARGSORT.

### First Bad Commit

Not a regression. The probe width is original to the feature (`6958d4136` /
`8544aba37`, #17004) and unchanged since.

### Relevant log output

<details>
<summary>Logs</summary>

```console
cfg: ngl=99 chain=4 n_vocab=151936 llama_set_sampler=true
llama_sampler_backend_support: device 'ROCm0' does not have support for op ARGSORT needed for sampler 'top-p'
ops: ROCm0    ne0=151936  SOFT_MAX=yes CUMSUM=yes SUM=yes ARGSORT=NO  TOP_K=yes
ops: ROCm0    ne0=1025    SOFT_MAX=yes CUMSUM=yes SUM=yes ARGSORT=NO  TOP_K=yes
ops: ROCm0    ne0=1024    SOFT_MAX=yes CUMSUM=yes SUM=yes ARGSORT=yes TOP_K=yes
ops: ROCm0    ne0=40      SOFT_MAX=yes CUMSUM=yes SUM=yes ARGSORT=yes TOP_K=yes
prompt: n_tokens=5 decode=0
tok 1: sample=    11 backend_token=    -1 cand_count=40     probs_count=0      top=[12095 1959 271]

# Vulkan, same position, same build tag
tok 1: sample=    11 backend_token=    11 cand_count=40     probs_count=40     top=[13:0.5266 11:0.4564 323:0.0170]

# CPU, same position
tok 1: sample=    11 backend_token=    11 cand_count=40     probs_count=40     top=[13:0.5167 11:0.4657 323:0.0176]
```
</details>

<details>
<summary>Build and run</summary>

```
curl -sLO https://github.com/ggml-org/llama.cpp/releases/download/b10798/llama-b10798-bin-ubuntu-rocm-10.0-x64.tar.gz
curl -sL -o src.tar.gz https://github.com/ggml-org/llama.cpp/archive/refs/tags/b10798.tar.gz
mkdir -p rocm src && tar xzf llama-b10798-bin-ubuntu-rocm-10.0-x64.tar.gz -C rocm && tar xzf src.tar.gz -C src

g++ -O2 -std=c++17 repro.cpp -o rocm/llama-b10798/repro \
  -I src/llama.cpp-b10798/include -I src/llama.cpp-b10798/ggml/include \
  -L rocm/llama-b10798 -lllama -lggml -lggml-base -Wl,-rpath,'$ORIGIN'

./rocm/llama-b10798/repro Qwen3-0.6B-Q8_0.gguf -ngl 99 -chain 4
```

The binary must sit in the release directory — `ggml_backend_load_all()`
searches the executable's own directory for the backend `.so` files.

</details>

<details>
<summary><code>repro.cpp</code></summary>

```cpp
#include "llama.h"
#include "ggml.h"
#include "ggml-backend.h"
#include <cstdio>
#include <cstdlib>
#include <cstring>

static void probe(int n_vocab) {
    ggml_init_params ip = { ggml_tensor_overhead()*512, NULL, true };
    ggml_context * g = ggml_init(ip);
    const int sizes[] = { n_vocab, 1025, 1024, 40 };
    for (size_t d = 0; d < ggml_backend_dev_count(); ++d) {
        ggml_backend_dev_t dev = ggml_backend_dev_get(d);
        for (int s : sizes) {
            ggml_tensor * a = ggml_new_tensor_2d(g, GGML_TYPE_F32, s, 1);
            ggml_tensor * ops[] = { ggml_soft_max(g, a), ggml_cumsum(g, a), ggml_sum(g, a),
                                    ggml_argsort(g, a, GGML_SORT_ORDER_DESC), ggml_top_k(g, a, s < 40 ? s : 40) };
            printf("ops: %-8s ne0=%-7d", ggml_backend_dev_name(dev), s);
            for (ggml_tensor * t : ops) printf(" %s=%s", ggml_op_name(t->op), ggml_backend_dev_supports_op(dev, t) ? "yes" : "NO ");
            printf("\n");
        }
    }
    ggml_free(g);
}

int main(int argc, char ** argv) {
    if (argc < 2) { fprintf(stderr, "usage: %s model.gguf [-ngl N] [-chain 1..5]\n", argv[0]); return 2; }
    int ngl = 99, chain = 4;
    for (int i = 2; i < argc; ++i) {
        if      (!strcmp(argv[i], "-ngl")   && i+1 < argc) ngl   = atoi(argv[++i]);
        else if (!strcmp(argv[i], "-chain") && i+1 < argc) chain = atoi(argv[++i]);
        else { fprintf(stderr, "bad arg %s\n", argv[i]); return 2; }
    }

    ggml_backend_load_all();
    llama_backend_init();

    llama_model_params mp = llama_model_default_params();
    mp.n_gpu_layers = ngl;
    llama_model * model = llama_model_load_from_file(argv[1], mp);
    if (!model) { printf("FAIL: model load\n"); return 2; }
    const llama_vocab * vocab = llama_model_get_vocab(model);
    const int n_vocab = llama_vocab_n_tokens(vocab);

    llama_sampler * smpl = llama_sampler_chain_init(llama_sampler_chain_default_params());
    if (chain >= 2)                llama_sampler_chain_add(smpl, llama_sampler_init_temp(1.0f));
    if (chain == 3 || chain == 4)  llama_sampler_chain_add(smpl, llama_sampler_init_top_k(40));
    if (chain == 4 || chain == 5)  llama_sampler_chain_add(smpl, llama_sampler_init_top_p(0.95f, 1));
    llama_sampler_chain_add(smpl, llama_sampler_init_dist(1234));

    llama_context_params cp = llama_context_default_params();
    cp.n_ctx = 512; cp.n_batch = 512; cp.n_ubatch = 512; cp.n_seq_max = 1; cp.no_perf = true;
    llama_context * ctx = llama_init_from_model(model, cp);
    if (!ctx) { printf("FAIL: context\n"); return 2; }

    printf("cfg: ngl=%d chain=%d n_vocab=%d llama_set_sampler=%s\n", ngl, chain, n_vocab,
           llama_set_sampler(ctx, 0, smpl) ? "true" : "false");
    probe(n_vocab);

    const char * prompt = "The capital of France is";
    llama_token toks[64];
    const int np = llama_tokenize(vocab, prompt, (int) strlen(prompt), toks, 64, true, true);
    if (np <= 0) { printf("FAIL: tokenize %d\n", np); return 2; }

    llama_batch b = llama_batch_init(512, 0, 1);
    auto add = [&](llama_token id, llama_pos p, bool l) {
        const int i = b.n_tokens;
        b.token[i] = id; b.pos[i] = p; b.n_seq_id[i] = 1; b.seq_id[i][0] = 0; b.logits[i] = l; b.n_tokens++;
    };

    b.n_tokens = 0;
    for (int i = 0; i < np; ++i) add(toks[i], i, i == np - 1);
    int rc = llama_decode(ctx, b);
    printf("prompt: n_tokens=%d decode=%d\n", np, rc);
    if (rc) { printf("FAIL: decode %d\n", rc); return 2; }

    int ith = np - 1;
    for (int g = 0; g < 8; ++g) {
        const llama_token   t    = llama_sampler_sample(smpl, ctx, -1);
        const uint32_t      cc   = llama_get_sampled_candidates_count_ith(ctx, ith);
        const uint32_t      pc   = llama_get_sampled_probs_count_ith(ctx, ith);
        const llama_token * cand = llama_get_sampled_candidates_ith(ctx, ith);
        const float       * pr   = llama_get_sampled_probs_ith(ctx, ith);

        printf("tok %d: sample=%6d backend_token=%6d cand_count=%-6u probs_count=%-6u",
               g, t, llama_get_sampled_token_ith(ctx, ith), cc, pc);
        if (cc && cand) {
            printf(" top=[");
            for (uint32_t i = 0; i < cc && i < 3; ++i) {
                printf("%s%d", i ? " " : "", cand[i]);
                if (pr && i < pc) printf(":%.4f", pr[i]);
            }
            printf("]");
        }
        printf("\n");

        b.n_tokens = 0; add(t, np + g, true);
        rc = llama_decode(ctx, b);
        if (rc) { printf("FAIL: decode %d\n", rc); break; }
        ith = 0;
    }

    llama_batch_free(b);
    llama_free(ctx);
    llama_sampler_free(smpl);
    llama_model_free(model);
    llama_backend_free();
    return 0;
}
```

</details>

The reproducer and the underlying investigation were AI-assisted.
